### PR TITLE
re-introduce implemented-requirements constraints

### PR DESCRIPTION
**Pull request by wandmagic**
# Committer Notes

re-introduce  two constraints for tracking implemented-requirement completion 
- extraneous-implemented-requirements
- incomplete-implemented-requirements

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/GSA/fedramp-automation/blob/master/CONTRIBUTING.md) guidance?
- [x] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/GSA/fedramp-automation/pulls) for the same update/change?
- [x] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] If applicable, have all [FedRAMP Documents Related to OSCAL Adoption](https://github.com/GSA/fedramp-automation) affected by the changes in this issue have been updated.?
- [ ] If applicable, does this PR reference the issue it addresses and explain how it addresses the issue?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### CODE REVIEW

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 1]
```diff
@@ -612,5 +612,41 @@
...
+                </message>
+            </index-has-key>
+            <index-has-key id='extraneous-implemented-requirements' target="//implemented-requirement" name="index-imported-controls" level="ERROR">
+                <formal-name>Additional Controls Implemented Not in Baseline</formal-name>
+                <description>A FedRAMP SSP MUST NOT include extraneous controls outside of the FedRAMP baseline.</description>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/general-concepts/implemented-requirements"/>
+                <key-field target="@control-id"/>
+                <message> A FedRAMP SSP MUST NOT include extraneous controls outside of the FedRAMP baseline
+                Extraneous ({@control-id})
+                </message>
```
Comment (Gabeblis) [inline src/validations/constraints/fedramp-external-constraints.xml:648]:
```suggestion
                <message>A FedRAMP SSP MUST NOT include extraneous controls outside of the FedRAMP baseline. Extraneous control: ({@control-id}).</message>
```

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 2]
```diff
@@ -612,5 +612,41 @@
...
+             <index name="index-implemented-requirements" target="$implemented-requirements-map">
+                <key-field target="@control-id"/>
+            </index>
+            <index-has-key id='incomplete-implemented-requirements' target="$imported-controls-map" name="index-implemented-requirements" level="ERROR">
+                <formal-name>Incomplete Implemented Requirements</formal-name>
+                <description>A FedRAMP SSP MUST contain an implemented requirement for each imported control.</description>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/general-concepts/implemented-requirements"/>
+                <key-field target="@id"/>
+                <message> A FedRAMP SSP MUST contain an implemented requirement for each imported control. Missing ({@id})
+                </message>
```
Comment (Gabeblis) [inline src/validations/constraints/fedramp-external-constraints.xml:639]:
```suggestion
                <message>A FedRAMP SSP MUST contain an implemented requirement for each imported control. Missing implemented requirement: ({@id}).</message>
```

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 3]
```diff
@@ -612,5 +612,41 @@
...
+        <metapath target="/catalog"/>
+        <metapath target="/profile"/>
+        <metapath target="/system-security-plan"/>
+        <constraints>
+         <let var="import-profile-href" expression="import-profile/@href"/>
+            <let var="resolved-import-profile-href" expression="if (starts-with($import-profile-href, '#')) then back-matter/resource[@uuid = substring($import-profile-href, 2)]/rlink/@href else $import-profile-href"/>
+            <let var="profile-is-available" expression="doc-available($resolved-import-profile-href)"/>
+            <let var="resolved-profile" expression="if ($profile-is-available) then resolve-profile(doc(resolve-uri($resolved-import-profile-href)))//catalog else //catalog"/>
+            <let var="imported-controls-map" expression="map:merge($resolved-profile//control ! map:entry(@id,.))?*"/>
+            <!-- Collect all the implemented requirements' control IDs -->
```
Comment (Gabeblis) [inline src/validations/constraints/fedramp-external-constraints.xml:625]:
```suggestion
```

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 4]
```diff
@@ -612,5 +612,41 @@
...
+            </index>
+            <index-has-key id='incomplete-implemented-requirements' target="$imported-controls-map" name="index-implemented-requirements" level="ERROR">
+                <formal-name>Incomplete Implemented Requirements</formal-name>
+                <description>A FedRAMP SSP MUST contain an implemented requirement for each imported control.</description>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/general-concepts/implemented-requirements"/>
+                <key-field target="@id"/>
+                <message> A FedRAMP SSP MUST contain an implemented requirement for each imported control. Missing ({@id})
+                </message>
+            </index-has-key>
+            <index-has-key id='extraneous-implemented-requirements' target="//implemented-requirement" name="index-imported-controls" level="ERROR">
```
Comment (Gabeblis) [inline src/validations/constraints/fedramp-external-constraints.xml:641]: I get that this does the exact same thing, but could we not reuse the map here for consistency? 
```suggestion
            <index-has-key id='extraneous-implemented-requirements' target="$implemented-requirements-map" name="index-imported-controls" level="ERROR">
```

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 5]
```diff
@@ -612,5 +612,41 @@
...
         </constraints>
     </context>
-
+<context>
+        <metapath target="/catalog"/>
+        <metapath target="/profile"/>
+        <metapath target="/system-security-plan"/>
+        <constraints>
+         <let var="import-profile-href" expression="import-profile/@href"/>
+            <let var="resolved-import-profile-href" expression="if (starts-with($import-profile-href, '#')) then back-matter/resource[@uuid = substring($import-profile-href, 2)]/rlink/@href else $import-profile-href"/>
```
Comment (Gabeblis) [inline src/validations/constraints/fedramp-external-constraints.xml:621]: Indentation looks a little off here.

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 6]
```diff
@@ -612,5 +612,41 @@
...
+            <let var="profile-is-available" expression="doc-available($resolved-import-profile-href)"/>
+            <let var="resolved-profile" expression="if ($profile-is-available) then resolve-profile(doc(resolve-uri($resolved-import-profile-href)))//catalog else //catalog"/>
+            <let var="imported-controls-map" expression="map:merge($resolved-profile//control ! map:entry(@id,.))?*"/>
+            <!-- Collect all the implemented requirements' control IDs -->
+            <let var="implemented-requirements-map" expression="map:merge(//implemented-requirement ! map:entry(@control-id,.))?*"/>
+            <index name="index-imported-controls" target="$imported-controls-map">
+                <key-field target="@id"/>
+            </index>
+             <index name="index-implemented-requirements" target="$implemented-requirements-map">
```
Comment (Gabeblis) [inline src/validations/constraints/fedramp-external-constraints.xml:630]: Indentation looks a little off here.

**Review by Gabeblis — CHANGES_REQUESTED** [review]
This is great stuff! Just some minor comments.

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 7]
```diff
@@ -612,5 +612,41 @@
...
+                <key-field target="@control-id"/>
+            </index>
+            <index-has-key id='incomplete-implemented-requirements' target="$imported-controls-map" name="index-implemented-requirements" level="ERROR">
+                <formal-name>Incomplete Implemented Requirements</formal-name>
+                <description>A FedRAMP SSP MUST contain an implemented requirement for each imported control.</description>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/general-concepts/implemented-requirements"/>
```
Comment (DimitriZhurkin) [inline src/validations/constraints/fedramp-external-constraints.xml:636]: For the `help-url` value, I'm getting the `404` error. However, I found `implemented-requirement` at this URL:
https://automate.fedramp.gov/documentation/ssp/6-security-controls/#control-definitions

Is documentation in progress?

File: src/validations/constraints/fedramp-external-constraints.xml (hunk 4 above)
Comment (wandmagic) [reply src/validations/constraints/fedramp-external-constraints.xml:641]: I'm not sure if we can re-use the map but i will try

File: src/validations/constraints/fedramp-external-constraints.xml (hunk 7 above)
Comment (wandmagic) [reply src/validations/constraints/fedramp-external-constraints.xml:636]: ah thank you i will update

**Comment by wandmagic** [comment]
Please note, i've added a style guide constraint change help URI is now tested with doc-available to assure that the help URI exists

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 8]
```diff
@@ -107,13 +112,34 @@
...
+            </index>
+            <index name="index-implemented-requirements" target="$implemented-requirements-map">
+                <key-field target="@control-id"/>
+            </index>
+            <index-has-key id='incomplete-implemented-requirements' target="$imported-controls-map" name="index-implemented-requirements" level="ERROR">
+                <formal-name>Incomplete Implemented Requirements</formal-name>
+                <description>A FedRAMP SSP MUST contain an implemented requirement for each imported control.</description>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/6-security-controls/#implementation-status"/>
+                <key-field target="@id"/>
+                <message> A FedRAMP SSP MUST contain an implemented requirement for each imported control. Missing ({@id})</message>
```
Comment (Gabeblis) [inline src/validations/constraints/fedramp-external-constraints.xml:132]:
```suggestion
                <message>A FedRAMP SSP MUST contain an implemented requirement for each imported control. Missing: ({@id}).</message>
```

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 9]
```diff
@@ -107,13 +112,34 @@
...
+            </index-has-key>
+            <index-has-key id='extraneous-implemented-requirements' target="//implemented-requirement" name="index-imported-controls" level="ERROR">
+                <formal-name>Additional Controls Implemented Not in Baseline</formal-name>
+                <description>A FedRAMP SSP MUST NOT include extraneous controls outside of the FedRAMP baseline.</description>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/6-security-controls/#implementation-status"/>
+                <key-field target="@control-id"/>
+                <message> A FedRAMP SSP MUST NOT include extraneous controls outside of the FedRAMP baseline Extraneous ({@control-id})</message>
```
Comment (Gabeblis) [inline src/validations/constraints/fedramp-external-constraints.xml:139]:
```suggestion
                <message>A FedRAMP SSP MUST NOT include extraneous controls outside of the FedRAMP baseline. Extraneous control: ({@control-id}).</message>
```

**Review by Gabeblis — DISMISSED** [review]
Thanks for addressing my feedback. Just a couple small tweaks to the messages that I'd like to see.

**Comment by wandmagic** [comment]
> Thanks for addressing my feedback. Just a couple small tweaks to the messages that I'd like to see.

applied, thanks for your dilligence

**Review by DimitriZhurkin — APPROVED** [review]

**Review by Gabeblis — APPROVED** [review]

**Review by aj-stein-gsa — COMMENTED** [review]
I know I am late to the party, but I did review it a little later in the day than planned. Nice work, @wandmagic.

**Pull request merged**